### PR TITLE
state machine

### PR DESCRIPTION
**Pull request by pangtao22**
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4)

### CODE REVIEW

File: src/plan_runner/plans/iiwa_plan_factory.cc [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  parser.AddModelFromFile(iiwa_path);
+  plant_->WeldFrames(plant_->world_frame(),
+                     plant_->GetFrameByName("iiwa_link_0"));
+  plant_->Finalize();
+}
+
+
+std::unique_ptr<PlanBase>
+IiwaPlanFactory::MakePlan(const drake::lcmt_robot_plan &msg_plan) const {
+  // TODO: replace this with a better test and use an lcm type with an
```
Comment (hjsuh94) [inline src/plan_runner/plans/iiwa_plan_factory.cc:29]: I agree :+1: we'll have to do this on our next round of PRs when Mark and I do new plans.

File: src/plan_runner/state_running.cc [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+PlanManagerStateBase* StateRunning::instance_ = nullptr;
+PlanManagerStateBase* StateRunning::Instance() {
+  if (!instance_) {
+    instance_ = new StateRunning();
+  }
+  return instance_;
+}
+
+const PlanBase * StateRunning::GetCurrentPlan(PlanManagerStateMachine *state_machine,
```
Comment (hjsuh94) [inline src/plan_runner/state_running.cc:16]: I understand the logic here (this is being called every 200Hz on the main thread), but the method name here a bit confusing, since it's doing a lot more than `GetCurrentPlan`, and is really in charge of starting, running, and stopping the plan (especially when in running state)

How about something like `HandleCurrentPlan`?

**Review by hjsuh94 — COMMENTED** [review]
Finished the first pass, and it looks awesome. I think the code is clean, and I double-checked that the FSM has the right flow of logic. Some high-level comments:

1. I think we need way more explicit documentation / commenting on things like the class definition. If you want I can take a stab at documenting everything on our next round of PRs. 
2. nit: I wonder if @mpetersen94 agrees, but the coding style was a bit difficult to parse coming from reading Drake code. (not a lot of blank lines between methods, etc.) - we can also reformat later if we decide to change the style. 
3. We might put all the files related to state_manager in a separate folder to make things cleaner :)

**Comment by pangtao22** [comment]
> 1. I think we need way more explicit documentation / commenting on things like the class definition. If you want I can take a stab at documenting everything on our next round of PRs.
> 2. nit: I wonder if @mpetersen94 agrees, but the coding style was a bit difficult to parse coming from reading Drake code. (not a lot of blank lines between methods, etc.) - we can also reformat later if we decide to change the style.
> 3. We might put all the files related to `state_manager` in a separate folder to make things cleaner :) (`plan_manager_state_manager` and all the `state_error`, `state_idle`, etc files)
> 4. Could we also include the client script in this PR?

2. The formatting is done with `clang-format`. There might be parameters that allow additional spaces between methods, etc.

File: src/plan_runner/iiwa_plan_manager.cc [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  // Compute command.
+  State s(Eigen::Map<VectorXd>(iiwa_status_msg_.joint_position_measured.data(),
+                               iiwa_status_msg_.num_joints),
+          Eigen::Map<VectorXd>(iiwa_status_msg_.joint_velocity_estimated.data(),
+                               iiwa_status_msg_.num_joints),
+          Eigen::Map<VectorXd>(iiwa_status_msg_.joint_torque_external.data(),
+                               iiwa_status_msg_.num_joints));
+  Command c;
+  if (plan) {
+    plan->Step(s, control_period_, state_machine_->GetCurrentPlanUpTime(t_now),
```
[the remark this answers is not in the document]
Comment (pangtao22) [reply src/plan_runner/iiwa_plan_manager.cc:133]: This thread is aways running and waiting for IIWA_STATUS messages. As soon as it gets one, the Iiwa status callback function (this function) is invoked to compute a command. Therefore, as long as computing the command doesn't take too long, commands are published at the same rate as the rate at which iiwa_status is being published.

File: src/plan_runner/iiwa_plan_manager.cc [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+#include "iiwa_plan_manager.h"
+#include "plans/plan_base.h"
+
+using Eigen::VectorXd;
+using std::cout;
+using std::endl;
+
+IiwaPlanManager::IiwaPlanManager(double control_period)
+    : control_period_(control_period) {
```
[the remark this answers is not in the document]
Comment (pangtao22) [reply src/plan_runner/iiwa_plan_manager.cc:14]: It will be used in `TaskSpacePlan` and others.

File: src/plan_runner/state_running.cc (hunk 2 above)
Comment (pangtao22) [reply src/plan_runner/state_running.cc]: I prefer to reserve names starting with Handle for lcm callbacks. What about `GetOrDiscardCurrentPlan`?

**Review by pangtao22 — COMMENTED** [review]
Didn't mean to start a review...

File: src/plan_runner/state_running.cc (hunk 2 above)
Comment (hjsuh94) [reply src/plan_runner/state_running.cc]: I think the right thing to do is to separate out the part that is starting and stopping the plan into a different method, to avoid it being overloaded with a method that returns the plan. 

But we can just merge and fix later :) I think it is a nit.

**Comment by pangtao22** [comment]
___
*[src/plan_runner/state_running.cc, line 16 at r1](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4#-MYS9PNNuS4GdeIrkNCR-r1-16:-MYS9j0c7w6IJLj4-dxI:b-7yd1q4) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/4af137831639a377182a537e7801f4255875ae27/src/plan_runner/state_running.cc#L16)):*
<details><summary><i>Previously, hjsuh94 (Hyung Ju Suh (Terry)) wrote…</i></summary><blockquote>

I think the right thing to do is to separate out the part that is starting and stopping the plan into a different method, to avoid it being overloaded with a method that returns the plan. 

But we can just merge and fix later :) I think it is a nit. 
</blockquote></details>

Added a TODO.

**Review by hjsuh94 — APPROVED** [review]
> Reviewed 13 of 31 files at r1, 14 of 15 files at r2, 6 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4)* status: all files reviewed, 2 unresolved discussions (waiting on @hjsuh94 and @mpetersen94)

**Review by mpetersen94 — COMMENTED** [review]
Sorry about the delay but I finally got around to reviewing this.  I know I missed the merge but thought I'd share my thoughts anyway.  At a high level I think this will work but there are some design decisions that I don't fully understand:

1. I don't love the state machine being split up as a series of classes with one for each state. I'm not sure what this gets us.  I think the thought is that we can extend this with more class types but the logic of switching between the states it tied up between them all and as a result the logic of the state machine feels hard to follow with it split up among 11 different files.  Am I missing the advantage?
2. I'm not sure I love locking the plan receiver such that it can only receive new plans when it is not running a plan.  I feel like that is more restrictive than necessary and means the only way to make the full stack reactive is to write a bespoke plan that is reactive (and deal with all the plumbing that is required) or use a streaming plan.  You can't send a new plan mid-execution to switch to.
3. Parsing the Cmake is a lot easier than most projects so I'm on board with the decision we've made.

Additional question not related to this PR, how did you set up CI and how does it run?  Would love to learn for my own edification.

> Reviewed 10 of 31 files at r1, 12 of 15 files at r2, 6 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4#-:-MYedC-yDZmSnxnPGCOQ:b543oxr)* status: all files reviewed, 5 unresolved discussions

___
*[src/plan_runner/iiwa_plan_manager.cc, line 133 at r1](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4#-MYS9PNNuS4GdeIrkNC5-r1-133:-MYefASs-Ac6kEVKk9qF:b-h91f5n) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/4af137831639a377182a537e7801f4255875ae27/src/plan_runner/iiwa_plan_manager.cc#L133)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

`void IiwaPlanManager::CalcCommandFromStatus()` is the function for the thread that computes commands, this function (`void IiwaPlanManager::HandleIiwaStatus`) is the lcm subscriber callback function for for IIWA_STATUS messages. As soon as an Iiwa status message is received, this function is invoked to compute a command and publishes it. Therefore, as long as computing the command doesn't take too long, commands are published at the same rate as the rate at which iiwa_status is being published.
</blockquote></details>

Putting the status and command in lockstep makes sense to me as well.
___
*[src/plan_runner/iiwa_plan_manager.cc, line 77 at r3](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4#-MYeemhg5u8VF8d808PK:-MYeemhg5u8VF8d808PL:b-ocn7as) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/3d4e3a49416a68d3df9a02b5afda0cb7b1083d10/src/plan_runner/iiwa_plan_manager.cc#L77)):*
> ```cpp
>       std::lock_guard<std::mutex> lock(mutex_state_machine_);
>       state_machine_->QueueNewPlan(std::move(plan));
>     }
> ```

It took me a while to parse that the mutex was being unlocked by garbage collection when the lock went out of scope.  It would be clearer if the lock was explicitly removed.
___
*[src/plan_runner/iiwa_plan_manager.cc, line 146 at r3](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4#-MYegCcv-wf5bwaZDHMb:-MYegCcv-wf5bwaZDHMc:b76jbay) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/3d4e3a49416a68d3df9a02b5afda0cb7b1083d10/src/plan_runner/iiwa_plan_manager.cc#L146)):*
> ```cpp
>     std::lock_guard<std::mutex> lock(mutex_state_machine_);
>     command_has_error = state_machine_->CommandHasError(s, c);
>   }
> ```

Same comment as above about locks
___
*[src/plan_runner/state_machine/plan_manager_state_machine.h, line 126 at r3](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4#-MYeb8Sa-ilvCeeHjkkK:-MYeb8Sa-ilvCeeHjkkL:ba40kc5) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/3d4e3a49416a68d3df9a02b5afda0cb7b1083d10/src/plan_runner/state_machine/plan_manager_state_machine.h#L126)):*
> ```objc
> };
> 
> inline bool PlanManagerStateMachine::has_received_status_msg() const {
> ```

What is the advantage of inlining the functions here instead of defining them above with the definition?
___
*[src/plan_runner/state_machine/state_running.cc, line 27 at r3](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4#-MYebu4J7J4wLz1YnCOk:-MYebu4KDVCi9E1uVraU:b-bpry4x) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/3d4e3a49416a68d3df9a02b5afda0cb7b1083d10/src/plan_runner/state_machine/state_running.cc#L27)):*
> ```cpp
>   } else {
>     const double t_elapsed_seconds = state_machine->GetCurrentPlanUpTime(t_now);
>     if (t_elapsed_seconds > plans.front()->duration() + 1.0) {
> ```

Why do we run the plan for one second past the duration of the plan?
___
*[src/plan_runner/state_machine/state_running.cc, line 53 at r3](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4#-MYec4mm9taNx4qM-bV-:-MYec4mm9taNx4qM-bV0:bacprv) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/3d4e3a49416a68d3df9a02b5afda0cb7b1083d10/src/plan_runner/state_machine/state_running.cc#L53)):*
> ```cpp
>                                 std::unique_ptr<PlanBase> plan) {
>   //TODO: what is the desired behavior when receiving a new plan while a plan
>   // is still being executed? Discard or queue?
> ```

Both spartan and the skeleton robot bridge that Siyuan pr'ed to drake a while ago replace the current plan whenever a new plan is received.  Spartan does have a ROS action workflow that allows the user to block until a plan is completed but the plan runner will switch plans whenever a new one is received.  This could be good behavior because it allows the user to change the executed plan mid-plan as new info is gathered.
___
*[src/plan_runner/state_machine/state_running.cc, line 69 at r3](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4#-MYecqHT6rSFCfHdMSis:-MYecqHT6rSFCfHdMSit:b-63rh68) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/3d4e3a49416a68d3df9a02b5afda0cb7b1083d10/src/plan_runner/state_machine/state_running.cc#L69)):*
> ```cpp
>       cmd.q_cmd.array().isNaN().sum() or cmd.tau_cmd.array().isNaN().sum();
> 
>   bool is_too_far_away = (state.q - cmd.q_cmd).norm() > 0.05;
> ```

Should this be a parameter or do we want this hard coded limit for all robots?

**Comment by pangtao22** [comment]
Hi Mark, just saw your comments.
> I don't love the state machine being split up as a series of classes with one for each state. I'm not sure what this gets us. I think the thought is that we can extend this with more class types but the logic of switching between the states it tied up between them all and as a result the logic of the state machine feels hard to follow with it split up among 11 different files. Am I missing the advantage?

Compared with a centralized blob of if-else statements, delegating the state machine's functions to different State objects allows us to unambiguously specify the behavior of each function in each state. Having state transitions defined in states instead of a centralized location also allows adding new states without altering existing ones. In fact this way of coding up a state machine is what software engineers call the "state design pattern", which is a new trick I learned from [this book](https://www.amazon.com/dp/0201633612/ref=cm_sw_em_r_mt_dp_VW0TJZ2E7BG0V2JY0GNW). 


> I'm not sure I love locking the plan receiver such that it can only receive new plans when it is not running a plan. I feel like that is more restrictive than necessary and means the only way to make the full stack reactive is to write a bespoke plan that is reactive (and deal with all the plumbing that is required) or use a streaming plan. You can't send a new plan mid-execution to switch to.

We can totally change this behavior if the use case for mid-execution switch arises. What's been implemented is just something simple that makes sense for the basic use cases I can think of. 

> Parsing the Cmake is a lot easier than most projects so I'm on board with the decision we've made.
Glad that it's been simple so far :D

**Review by pangtao22 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4)* status: all files reviewed, 5 unresolved discussions

___
*[src/plan_runner/iiwa_plan_manager.cc, line 133 at r1](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4#-MYS9PNNuS4GdeIrkNC5-r1-133:-MYl9r2T7wIJVmo8nH32:b-n3ri58) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/4af137831639a377182a537e7801f4255875ae27/src/plan_runner/iiwa_plan_manager.cc#L133)):*
<details><summary><i>Previously, mpetersen94 (Mark Petersen) wrote…</i></summary><blockquote>

Putting the status and command in lockstep makes sense to me as well.
</blockquote></details>

But we don't have to? The Step function does not affect the state machine, but it is probably the most time-consuming part of this thread. I think it is better to block other threads as shortly as possible, although I don't know how much overhead `lock_guard` adds.
___
*[src/plan_runner/state_machine/plan_manager_state_machine.h, line 126 at r3](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4#-MYeb8Sa-ilvCeeHjkkK:-MYl9J-E-neWNsFRP2eK:b48rljv) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/3d4e3a49416a68d3df9a02b5afda0cb7b1083d10/src/plan_runner/state_machine/plan_manager_state_machine.h#L126)):*
<details><summary><i>Previously, mpetersen94 (Mark Petersen) wrote…</i></summary><blockquote>

What is the advantage of inlining the functions here instead of defining them above with the definition?
</blockquote></details>

Methods defined in the class definition are automatically inlined. These ones are defined outside the class body so that the class definition doesn't look too bloated. I admit that the decision between defining an inline function up there or down here is a bit arbitrary.
___
*[src/plan_runner/state_machine/state_running.cc, line 27 at r3](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4#-MYebu4J7J4wLz1YnCOk:-MYlAXdv6g79y_IAy75U:b-9389xj) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/3d4e3a49416a68d3df9a02b5afda0cb7b1083d10/src/plan_runner/state_machine/state_running.cc#L27)):*
<details><summary><i>Previously, mpetersen94 (Mark Petersen) wrote…</i></summary><blockquote>

Why do we run the plan for one second past the duration of the plan?
</blockquote></details>

It's mean't to let the plan "stabilize", which in practice seems to make transitions between plans smoother. IIWA's internal controller does some low-pass filtering on the commands it receives, so the command we send at the last tick in the plan's duration is executed on IIWA a couple of milliseconds later. In addition, some reference trajectories, such as PiecewiseQuaternionSlerp, has non-zero reference velocity near the end of the plan. Therefore giving it some extra time to stabilize to zero velocity also seems to reduce jittering between plan switchings, based on what I remember from 6.881 labs. 
___
*[src/plan_runner/state_machine/state_running.cc, line 53 at r3](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4#-MYec4mm9taNx4qM-bV-:-MYlBCD37A-i7aN2wRSY:buvn884) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/3d4e3a49416a68d3df9a02b5afda0cb7b1083d10/src/plan_runner/state_machine/state_running.cc#L53)):*
<details><summary><i>Previously, mpetersen94 (Mark Petersen) wrote…</i></summary><blockquote>

Both spartan and the skeleton robot bridge that Siyuan pr'ed to drake a while ago replace the current plan whenever a new plan is received.  Spartan does have a ROS action workflow that allows the user to block until a plan is completed but the plan runner will switch plans whenever a new one is received.  This could be good behavior because it allows the user to change the executed plan mid-plan as new info is gathered.
</blockquote></details>

Agreed. Maybe our implementation can support both, and ask the user to specify the desired behavior in a configuration file? It shouldn't be too hard to add when we need such a feature. 
___
*[src/plan_runner/state_machine/state_running.cc, line 69 at r3](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/4#-MYecqHT6rSFCfHdMSis:-MYlAkjN9_RuHb5nNW1y:brclmer) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/3d4e3a49416a68d3df9a02b5afda0cb7b1083d10/src/plan_runner/state_machine/state_running.cc#L69)):*
<details><summary><i>Previously, mpetersen94 (Mark Petersen) wrote…</i></summary><blockquote>

Should this be a parameter or do we want this hard coded limit for all robots?
</blockquote></details>

Exactly. There are a few hardcoded parameters such as sdf file path to the robot, controller period and the "extra time" after each plan. We will move them into a yaml file in the future.
